### DISCUSSION TITLE
Behavior lifecycle function onCreated called early

### DISCUSSION THREAD

**Discussion by arthuro555** [4ian/GDevelop#3634] [Enhancement ideas (and their technical discussions)]
## Description

Not sure if this should be considered a bug, but the behavior lifecycle `onCreated` function is called before the object is completely initialized. For example, it does not yet contain its creation position. A workaround for this is to put an event with a trigger once in the pre-/postevents lifecycle functions but this is not viable since if an object is created in events, the behavior will be in an invalid state until the end of the events is reached (for the lifecycle functions to be called).

## To Reproduce

1. Open this game: [game.zip](https://github.com/4ian/GDevelop/files/7148289/game.zip)
2. See that the event is in onCreated and how the default position is not set on a preview
3. Enable the event in onPreEvents and see how it now works

**Comment by 4ian** [comment]
This is because `extraInitializationFromInitialInstance` or setPosition/setLayer/setZOrder in `doCreateObjectOnScene` are called after the constructor.
We could have a new lifecycle event that would replace onCreated that would be called after these.

**Comment by arthuro555** [comment]
That is not a feature request, that's a bug report, the bug being that object data is not accessible (whether read or write) in that function while it could and should. This shouldn't have been moved into a discussion.

**Reply by 4ian** [reply]
The object data is accessible though - it's the position and any instance specific initialisation that is not done yet?

**Reply by arthuro555** [reply]
That is what I meant by object data, excuse me, I should have used the correct term. You cannot use or set the position, angle, etc of an object which makes the onCreated useless in most cases.

**Reply by 4ian** [reply]
Ok, we're aligned on the issue. We might need a "onInitialized" lifecycle then. The "onCreated" is valid in the sense that the object is constructed (as in "the constructor is finished"). It might still has some usefuless (set up some default properties or stuff), but I guess most of the time you want to run the behavior code in "onInitialized". Note that the game engine will be responsible to run this (while the onCreated is run by the object class itself).

**Comment by Bouh** [comment]
Converted into issue on https://github.com/4ian/GDevelop/issues/3663
